### PR TITLE
statistics: clean up dropped predicate columns stats usage

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/53567

Problem Summary:

We need to clean up the outdated predicate columns that have been dropped from the schema.
Otherwise, we will attempt to analyze non-existent columns.

### What changed and how does it work?

1. Added a test case to verify current behavior.
2. Added a new function to clean up the non-existent columns before we get the predicate columns.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

**Comment by hi-rustin** [comment]
Tested locally:
1. Start the TiDB cluster with patch: `tiup playground nightly  --db.binpath /Users/hi-rustin/vsc/tidb/bin/tidb-server`
2. Create a new table:
```sql
mysql> create table t (a int, b int);
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t values (1, 1), (2, 2), (3, 3);
Query OK, 3 rows affected (0.01 sec)
Records: 3  Duplicates: 0  Warnings: 0
```
3. Enable predicate column tracking:
```sql
mysql> set global tidb_enable_column_tracking = 1;
Query OK, 0 rows affected (0.02 sec)

mysql> select @@tidb_enable_column_tracking;
+-------------------------------+
| @@tidb_enable_column_tracking |
+-------------------------------+
|                             1 |
+-------------------------------+
1 row in set (0.00 sec)
```
4. Select data with predicates:
```sql
mysql> select * from t where b > 1;
+------+------+
| a    | b    |
+------+------+
|    2 |    2 |
|    3 |    3 |
+------+------+
2 rows in set (0.00 sec)
```
5. Wait for 2 minutes:
```sql
mysql> SELECT * FROM MYSQL.COLUMN_STATS_USAGE;
+----------+-----------+---------------------+------------------+
| table_id | column_id | last_used_at        | last_analyzed_at |
+----------+-----------+---------------------+------------------+
|      104 |         2 | 2024-05-30 16:47:18 | NULL             |
+----------+-----------+---------------------+------------------+
1 row in set (0.01 sec)
```
6. Analyze with predicate columns: `ANALYZE test.t PREDICATE COLUMNS`
```sql
mysql> ANALYZE TABLE test.t PREDICATE COLUMNS;
Query OK, 0 rows affected, 1 warning (0.03 sec)
mysql> select * from mysql.analyze_options;
+----------+------------+-------------+---------+------+---------------+------------+
| table_id | sample_num | sample_rate | buckets | topn | column_choice | column_ids |
+----------+------------+-------------+---------+------+---------------+------------+
|      104 |          0 |           0 |       0 |   -1 | PREDICATE     |            |
+----------+------------+-------------+---------+------+---------------+------------+
1 row in set (0.00 sec)
```
7. Check analyze jobs:
```sql
mysql> SELECT * FROM mysql.analyze_jobs;
+----+---------------------+--------------+------------+----------------+------------------------------------------------------------------+----------------+---------------------+---------------------+----------+-------------+----------------+------------+
| id | update_time         | table_schema | table_name | partition_name | job_info                                                         | processed_rows | start_time          | end_time            | state    | fail_reason | instance       | process_id |
+----+---------------------+--------------+------------+----------------+------------------------------------------------------------------+----------------+---------------------+---------------------+----------+-------------+----------------+------------+
|  1 | 2024-05-30 16:52:28 | test         | t          |                | analyze table columns b with 256 buckets, 500 topn, 1 samplerate |              3 | 2024-05-30 16:52:28 | 2024-05-30 16:52:28 | finished | NULL        | 127.0.0.1:4000 |       NULL |
+----+---------------------+--------------+------------+----------------+------------------------------------------------------------------+----------------+---------------------+---------------------+----------+-------------+----------------+------------+
1 row in set (0.00 sec)
```
8. Drop a column:
```sql
mysql> alter table t drop column b;
Query OK, 0 rows affected (0.10 sec)
```
7. Check the stats usage:
```sql
mysql> SELECT * FROM MYSQL.COLUMN_STATS_USAGE;
+----------+-----------+---------------------+---------------------+
| table_id | column_id | last_used_at        | last_analyzed_at    |
+----------+-----------+---------------------+---------------------+
|      104 |         2 | 2024-05-30 16:47:18 | 2024-05-30 16:52:28 |
+----------+-----------+---------------------+---------------------+
1 row in set (0.00 sec)
```
8. Try analyze again:
```sql
mysql> ANALYZE TABLE test.t PREDICATE COLUMNS;
Query OK, 0 rows affected, 1 warning (0.02 sec)
mysql> show warnings;
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                 |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1105 | No predicate column has been collected yet for table test.t so all columns are analyzed                                                 |
| Note    | 1105 | Analyze use auto adjusted sample rate 1.000000 for table test.t, reason to use this rate is "use min(1, 110000/3) as the sample-rate=1" |
+---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
2 rows in set (0.00 sec)
```
8. Check the jobs again:
```sql
mysql> SELECT * FROM mysql.analyze_jobs;
+----+---------------------+--------------+------------+----------------+--------------------------------------------------------------------+----------------+---------------------+---------------------+----------+-------------+----------------+------------+
| id | update_time         | table_schema | table_name | partition_name | job_info                                                           | processed_rows | start_time          | end_time            | state    | fail_reason | instance       | process_id |
+----+---------------------+--------------+------------+----------------+--------------------------------------------------------------------+----------------+---------------------+---------------------+----------+-------------+----------------+------------+
|  1 | 2024-05-30 16:52:28 | test         | t          |                | analyze table columns b with 256 buckets, 500 topn, 1 samplerate   |              3 | 2024-05-30 16:52:28 | 2024-05-30 16:52:28 | finished | NULL        | 127.0.0.1:4000 |       NULL |
|  2 | 2024-05-30 16:53:21 | test         | t          |                | analyze table all columns with 256 buckets, 500 topn, 1 samplerate |              3 | 2024-05-30 16:53:21 | 2024-05-30 16:53:21 | finished | NULL        | 127.0.0.1:4000 |       NULL |
+----+---------------------+--------------+------------+----------------+--------------------------------------------------------------------+----------------+---------------------+---------------------+----------+-------------+----------------+------------+
2 rows in set (0.00 sec)
```
9. Check stats usage again:
```sql
mysql> SELECT * FROM MYSQL.COLUMN_STATS_USAGE;
+----------+-----------+--------------+---------------------+
| table_id | column_id | last_used_at | last_analyzed_at    |
+----------+-----------+--------------+---------------------+
|      104 |         1 | NULL         | 2024-05-30 16:53:21 |
+----------+-----------+--------------+---------------------+
1 row in set (0.00 sec)
```

**I am unsure why we insert the column 'a' into the MYSQL.COLUMN_STATS_USAGE table. I don't think it makes sense, but it is off-topic for this PR. We will fix it later.**

**Comment by hi-rustin** [comment]
> I am unsure why we insert the column 'a' into the MYSQL.COLUMN_STATS_USAGE table. I don't think it makes sense, but it is off-topic for this PR. We will fix it later.

I don't if this is a bug or a feature.

But the minimal reproduction steps are:
```sql
mysql> use test;
Database changed
mysql> create table t (a int, b int);
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t values (1, 1), (2, 2), (3, 3);
Query OK, 3 rows affected (0.00 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> set global tidb_enable_column_tracking = 1;
Query OK, 0 rows affected (0.01 sec)

mysql> analyze table t;
Query OK, 0 rows affected, 1 warning (0.03 sec)

mysql> SELECT * FROM MYSQL.COLUMN_STATS_USAGE;
+----------+-----------+--------------+---------------------+
| table_id | column_id | last_used_at | last_analyzed_at    |
+----------+-----------+--------------+---------------------+
|      104 |         1 | NULL         | 2024-05-30 17:00:36 |
|      104 |         2 | NULL         | 2024-05-30 17:00:36 |
+----------+-----------+--------------+---------------------+
2 rows in set (0.00 sec)
```

I will figure out why this happens.

**Review by qw4990 — APPROVED** [review]

File: pkg/statistics/handle/usage/predicate_column.go [hunk 1]
```diff
@@ -159,6 +166,33 @@ func GetPredicateColumns(sctx sessionctx.Context, tableID int64) ([]int64, error
...
+	}
+	allColumns := table.Meta().Columns
+	// Due to SQL limitations, column IDs must be converted to strings for proper escaping in the query :(
+	columnIDs := make([]string, 0, len(allColumns))
+	for _, col := range allColumns {
+		columnIDs = append(columnIDs, fmt.Sprintf("%d", col.ID))
+	}
+
+	// Delete the column stats usage information whose column is dropped.
+	_, _, err := utilstats.ExecRows(
```
Comment (hawkingrei) [inline pkg/statistics/handle/usage/predicate_column.go:186]: Do we need to add ```ActionDropColumn```  event to delete the column stats usage information ?

**Review by hawkingrei — APPROVED** [review]

File: pkg/statistics/handle/usage/predicate_column.go (hunk 1 above)
Comment (AilinKid) [reply pkg/statistics/handle/usage/predicate_column.go:186]: make sense

Comment (hi-rustin) [reply pkg/statistics/handle/usage/predicate_column.go:186]: img_v3_02bd_3a5aecb0-4bb4-49e7-927d-8edd1438103g
There is a reason why we didn't use the DDL event.

**Pull request merged**